### PR TITLE
Top and bottom line display depending on the position of the event in…

### PR DESCRIPTION
**Pull request by RobinCsl**
… the timeline

### CODE REVIEW

File: app/MMB/src/scenes/timeline/Timeline.js [hunk 1]
```diff
@@ -53,35 +54,53 @@ function getValidTimelineEvent(data) {
...
+    const prevTimestamp = idx(prevRenderedEvent, _ => _.props.data.timestamp);
+    if (timestamp && prevTimestamp) {
+      const eventTimestamp = new Date(timestamp);
+      const prevEventTimestamp = new Date(prevTimestamp);
+      if (!DateUtils.isSameDay(eventTimestamp, prevEventTimestamp)) {
+        return <DaySeparator date={eventTimestamp} />;
+      }
     }
   }
   return null;
```
Comment (tbergq) [inline app/MMB/src/scenes/timeline/Timeline.js]: Could you please try to refactor this function a bit? The level of nested if/else makes it hard to read and understand.

Comment (RobinCsl) [reply app/MMB/src/scenes/timeline/Timeline.js]: Ok, will do.

File: app/MMB/src/scenes/timeline/Timeline.js [hunk 2]
```diff
@@ -53,35 +54,53 @@ function getValidTimelineEvent(data) {
...
-          daySeparator = renderDaySeparator(event, events[index - 1]);
-        }
-        return (
-          <React.Fragment key={'TimelineEvent-' + event.__typename + index}>
-            {daySeparator}
-            {getValidTimelineEvent(event)}
-          </React.Fragment>
+    const renderedEvents = events
+      .map(event => getValidTimelineEvent(event))
+      .filter(renderedEvent => renderedEvent);
```
Comment (tbergq) [inline app/MMB/src/scenes/timeline/Timeline.js]: Why is this needed?

File: app/MMB/src/scenes/timeline/TimelineEventIcon.js [hunk 3]
```diff
@@ -4,9 +4,13 @@ import * as React from 'react';
 import { View } from 'react-native';
 import { StyleSheet, Color, Icon, TextIcon } from '@kiwicom/mobile-shared';
 
+export const TimelineEventIconContext = React.createContext({
+  displayTopLine: true,
+  displayBottomLine: true,
+});
+
```
Comment (tbergq) [inline app/MMB/src/scenes/timeline/TimelineEventIcon.js]: What is this context used for? 
You set two booleans, why can it not be variables?

File: app/MMB/src/scenes/timeline/Timeline.js (hunk 2 above)
Comment (RobinCsl) [reply app/MMB/src/scenes/timeline/Timeline.js]: There could be some additional events given by the GraphQL server which would not correspond to any implemented event component.
There could be situations where that adds an additional day banner when no component is being rendered, as `getValidTimelineEvent` would return `null`; also, since the first and last events need to display the vertical bars in a different way, the length of the non-null components must be taken into account, not the length of the events.
Does that make sense?

File: app/MMB/src/scenes/timeline/TimelineEventIcon.js (hunk 3 above)
Comment (RobinCsl) [reply app/MMB/src/scenes/timeline/TimelineEventIcon.js]: This is to indicate default values. But that context is set depending on the index of the rendered event components.

File: app/MMB/src/scenes/timeline/Timeline.js (hunk 2 above)
Comment (tbergq) [reply app/MMB/src/scenes/timeline/Timeline.js]: Yes

Comment (tbergq) [reply app/MMB/src/scenes/timeline/Timeline.js]: Maybe you could add a comment explaining it

File: app/MMB/src/scenes/timeline/Timeline.js (hunk 1 above)
Comment (RobinCsl) [reply app/MMB/src/scenes/timeline/Timeline.js]: Done

**Comment by RobinCsl** [comment]
We will need to revisit DateUtils and DateFormatter.
This is the current state of the timeline for one journey:
simulator screen shot - iphone 6 - 2018-06-27 at 18 23 36

And this is a screenshot from the currently deployed app:
img_0274

We can see the dates and times are off.

**Comment by tbergq** [comment]
> We will need to revisit DateUtils and DateFormatter.

I think our DateFormatter and DateUtils are wrong. 

If I fly from Barcelona - Lima with return flight, and we format the flight time in UTC it will give me the wrong time. I am interested in seeing the actual time the flight takes off in that timezone.

@mrtnzlml please correct me if I am wrong.

File: app/MMB/src/scenes/timeline/Timeline.js [hunk 4]
```diff
@@ -53,35 +54,47 @@ function getValidTimelineEvent(data) {
...
+            timestamp &&
+            DateUtils.isSameDay(
+              new Date(currentTimestamp),
+              new Date(timestamp),
+            ),
         );
+        if (index === sameDays.indexOf(true)) {
+          daySeparator = <DaySeparator date={new Date(currentTimestamp)} />;
+        }
       }
```
Comment (tbergq) [inline app/MMB/src/scenes/timeline/Timeline.js]: It seems like you map every event to get the type, then loop to remove nulls, then loop to get all timestamps. Then loop over the events, and for each event loop the timestamps.

Would it be possible here to use reduce, loop once and return an object.

```
events.reduce((acc, curr) => {
const event = getValidTimelineEvent(curr);
if (!event) return acc;
// attach curr to acc and return
}, {})
```

having the structure like {'2018-06-01': { events: [...] }, '2018-06-02': { events : [...]}

Then loop over the keys in the object and between each key add the day separator.

Do you think this makes sense?

Comment (RobinCsl) [reply app/MMB/src/scenes/timeline/Timeline.js]: Yes, that's a much cleaner way of doing it. I will give it a try. Thank you!

**Comment by mrtnzlml** [comment]
@tbergq I don't know man. Are you sure you are not doing something dodgy there? We should always get the date in the UTC format (at least from our GraphQL):

```json
{
  "data": {
    "allFlights": {
      "edges": [
        {
          "node": {
            "arrival": {
              "localTime": "2018-08-10T00:40:00.000Z",
              "time": "2018-08-10T05:40:00.000Z"
            },
            "departure": {
              "localTime": "2018-08-08T23:55:00.000Z",
              "time": "2018-08-08T21:55:00.000Z"
            }
          }
        },
...
```

It's defined by the `DateTime` type in GraphQL:

> A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the date-time format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.

So this is why it's so important to call the Intl with UTC timezone (so it will stay in UTC and it won't be converted into local timezone):

```
Intl.DateTimeFormat('en', {timeZone: 'UTC', year: 'numeric', month: 'numeric', day: 'numeric', hour: 'numeric', minute: 'numeric'}).format(new Date('2018-08-10T00:40:00.000Z'))
> "8/10/2018, 12:40 AM"   (correct)

BUT

Intl.DateTimeFormat('en', {year: 'numeric', month: 'numeric', day: 'numeric', hour: 'numeric', minute: 'numeric'}).format(new Date('2018-08-10T00:40:00.000Z'))
> "8/10/2018, 2:40 AM"   (WTF? It's not true...)
```

Maybe I don't understand what it's wrong exactly. Can you please elaborate it more with some example?
